### ISSUE TITLE
Support for many cuda compute capabilities

### ISSUE THREAD

**Issue by valassi** [madgraph5/madgraph4gpu#533]
Hi @hageboeck this is a followup to your recent changes (which made a lot of sense!) adding an env variable MADGRAPH_CUDA_ARCHITECTURE to decide which compute capability to build for
https://github.com/madgraph5/madgraph4gpu/commit/058144cc6888ee5aa6c1689a224797380abbc0c9

I realised that there are use cases where we need to build the code for more than a single compute architecture
- One clear use case is what I do in the benchmarking containers: I build a docker container once, that may later land on any Nvidia card, including P100, V100, A100, etc. For the moment my "v0.6" container uses V100 (capability 70), and I tested it on V100s but not elsewhere, so I assume it would break. (A different story is that I used cuda11.7 with the latest driver, and it does not run on machines with older drivers installed, but there the only solution I see is that nodes should always have the latest drivers).
- Another possible use case is software builds for distribution eg via cvmfs, as we do routinely. This may not be a strong constraint however, as each process needs to be separately regenerated and rebuilt. But maybe the experments do want to build once their gridpacks and then send them to different GPUs...
- Then there are cases like an HPC center (or even our CERN centre) where you may want to build the code once on login nodes and then run it on possible different partitions of the data centre, using different GPUs.

There are various options here
1. The simplest would be to modify the logic of  MADGRAPH_CUDA_ARCHITECTURE, for instance so that "70,80" means "build both 70 and 80. This in principle is straightforward, but doing a loop in Makefile language over an array of comma separated variables may be a bit cumbersome. This produces fat binaries supporting both 70 and 80.
2. A similar one, that is simpler to code but more complex for users would be to allow specifying arbitrary cuda flags from outside. But this becomes much more cumbersome than specifying MADGRAPH_CUDA_ARCHITECTURE=70,80. This also produces fat binaries supporting both 70 and 80.
3. A very different approach would consist in having separate build directories for cuda70 and cuda80. This produces separate thin binaries supporting either 70 or 80. This is something to keep in mind, because it is also related to many other changes that eventually will come. For instance, we may need separate fortran builds for cuda and for c++. And we already have separate simd builds in c++, which eventually I would like to convert instead to fat binaries. One thing that I could do here is to add the compute capability to the "build tag" I use (which now includes the simd mode).

Any suggestion? I would probably go for option 1 initially, and defer any more radical choices to a more comprehensive rethink of fat binaries later on.

By the way it would be interesting to know what our colleagues in SFT are doing when building cuda libraries (provided they have to build any at all...?). It may be worth asking them (I can send an email to LIM team, or you can ask informally?).

Finally one question, which is probably related. The code for 70 used to be
```
CUARCHFLAGS = -gencode arch=compute_70,code=compute_70 -gencode arch=compute_70,code=sm_70
```
You now changed it to
```
CUARCHFLAGS = --gpu-architecture=compute_70 --gpu-code=sm_70,compute_70
```
I am totally not an expert, are these equivalent, just using different syntaxes? Or are they doing different things?

Thanks!
Andrea

**Comment by valassi** [comment]
For what concerns benchmark containers I have opened the corresponding ticket https://its.cern.ch/jira/browse/BMK-1060

**Comment by valassi** [comment]
I have added a fix using option 1,
https://github.com/madgraph5/madgraph4gpu/commit/b359bac9c19ed8ecbd7761f676f7ff89fb49e67e

Note that I had to go back to my previous use of nvcc options.
First attempt with current options failed
https://github.com/madgraph5/madgraph4gpu/commit/ba336add2725a7f60c75d67a42d49a4f321d590d
```
MADGRAPH_CUDA_ARCHITECTURE=60,70,80 make
fails with

ccache /usr/local/cuda-11.7/bin/nvcc  -O3  -lineinfo -I. -I../../src -I../../../../../tools -I/usr/local/cuda-11.7/include/ -DUSE_NVTX --gpu-architecture=compute_60 --gpu-code=sm_60,compute_60 --gpu-architecture=compute_70 --gpu-code=sm_70,compute_70 --gpu-architecture=compute_80 --gpu-code=sm_80,compute_80 -use_fast_math -std=c++17  -ccbin /cvmfs/sft.cern.ch/lcg/releases/gcc/11.2.0-ad950/x86_64-centos7/bin/g++ -DMGONGPU_FPTYPE_DOUBLE -Xcompiler -fPIC -c gcheck_sa.cu -o gcheck_sa.o
nvcc warning : incompatible redefinition for option 'gpu-architecture', the last value of this option was used
nvcc warning : incompatible redefinition for option 'gpu-architecture', the last value of this option was used
nvcc fatal   : Incompatible code generation requested: arch='compute_80', code='sm_60'
```
Second attempt with older options succeeded
https://github.com/madgraph5/madgraph4gpu/commit/fd29c23ce6fd200a94dff776d7a7bb4e7f0d49fe
```
MADGRAPH_CUDA_ARCHITECTURE=60,70,80 make
succeeds:
ccache /usr/local/cuda-11.7/bin/nvcc  -O3  -lineinfo -I. -I../../src -I../../../../../tools -I/usr/local/cuda-11.7/include/ -DUSE_NVTX -gencode arch=compute_60,code=compute_60 -gencode arch=compute_60,code=sm_60 -gencode arch=compute_70,code=compute_70 -gencode arch=compute_70,code=sm_70 -gencode arch=compute_80,code=compute_80 -gencode arch=compute_80,code=sm_80 -use_fast_math -std=c++17  -ccbin /cvmfs/sft.cern.ch/lcg/releases/gcc/11.2.0-ad950/x86_64-centos7/bin/g++ -DMGONGPU_FPTYPE_DOUBLE -Xcompiler -fPIC -c gcheck_sa.cu -o gcheck_sa.o
```

I keep this open for the moment as more discussion and other options may be useful

**Comment by valassi** [comment]
This is fixed in MR #532 - I link it there so it will get closed automatically. No need to review further.
